File: cache/src/main/java/com/mm/http/RealRequestInterceptorChain.java
package com.mm.http;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.mm.http.cache.CacheHelper;

import java.io.IOException;
import java.net.HttpURLConnection;
import java.util.List;
import java.util.concurrent.TimeUnit;

import okhttp3.Call;
import okhttp3.Connection;
import okhttp3.Interceptor;
import okhttp3.Protocol;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.internal.Util;

/**
 * Created by : majian
 * Date : 2021/8/25
 * 处理自定义拦截器中对request 请求更改的逻辑。
 */

public class RealRequestInterceptorChain implements Interceptor.Chain {
    private final Call call;
    private final List<Interceptor> interceptors;
    private final int index;
    private final Request request;
    private final int connectTimeoutMillis;
    private final int readTimeoutMillis;
    private final int writeTimeoutMillis;
    private final CacheHelper cache;

    RealRequestInterceptorChain(Call call, CacheHelper cache, List<Interceptor> interceptors, int index, Request request, int connectTimeoutMillis, int readTimeoutMillis, int writeTimeoutMillis) {
        this.interceptors = interceptors;
        this.index = index;
        this.request = request;
        this.call = call;
        this.connectTimeoutMillis = connectTimeoutMillis;
        this.readTimeoutMillis = readTimeoutMillis;
        this.writeTimeoutMillis = writeTimeoutMillis;
        this.cache = cache;
    }

    @NonNull
    @Override
    public Call call() {
        return call;
    }

    @Override
    public int connectTimeoutMillis() {
        return connectTimeoutMillis;
    }

    @Nullable
    @Override
    public Connection connection() {
        return null;
    }

    @NonNull
    @Override
    public Response proceed(@NonNull Request request) throws IOException {
        // Call the next interceptor in the chain.
        if (index < interceptors.size()) {
            RealRequestInterceptorChain next = copy(index + 1, request, connectTimeoutMillis, readTimeoutMillis, writeTimeoutMillis);
            Interceptor interceptor = interceptors.get(index);
            return interceptor.intercept(next);
        }
        Response response = cache.get(request);
        if (response != null) {
            return response.newBuilder().request(request).build();
        }
        return new Response.Builder().request(request)
                .protocol(Protocol.HTTP_1_1).code(HttpURLConnection.HTTP_GATEWAY_TIMEOUT)
                .message("Unsatisfiable Request (only-if-cached)")
                .body(Util.EMPTY_RESPONSE).build();
    }

    @Override
    public int readTimeoutMillis() {
        return readTimeoutMillis;
    }

    @NonNull
    @Override
    public Request request() {
        return request;
    }

    @NonNull
    @Override
    public Interceptor.Chain withConnectTimeout(int i, @NonNull TimeUnit timeUnit) {
        int connectTimeoutMillis = checkDuration(i, timeUnit);
        return copy(index, request, connectTimeoutMillis, readTimeoutMillis, writeTimeoutMillis);
    }

    @NonNull
    @Override
    public Interceptor.Chain withReadTimeout(int i, @NonNull TimeUnit timeUnit) {
        int readTimeoutMillis = checkDuration(i, timeUnit);
        return copy(index, request, connectTimeoutMillis, readTimeoutMillis, writeTimeoutMillis);
    }

    @NonNull
    @Override
    public Interceptor.Chain withWriteTimeout(int i, @NonNull TimeUnit timeUnit) {
        int writeTimeoutMillis = checkDuration(i, timeUnit);
        return copy(index, request, connectTimeoutMillis, readTimeoutMillis, writeTimeoutMillis);
    }

    private RealRequestInterceptorChain copy(int index, Request request,
                                             int connectTimeoutMillis, int readTimeoutMillis, int writeTimeoutMillis) {
        return new RealRequestInterceptorChain(call, cache, interceptors, index, request, connectTimeoutMillis, readTimeoutMillis, writeTimeoutMillis);
    }

    private int checkDuration(int duration, TimeUnit unit) {
        if (duration < 0) {
            throw new IllegalArgumentException("duration < 0");
        }
        if (unit == null) {
            throw new IllegalArgumentException("unit == null");
        }
        return (int) unit.toMillis(duration);
    }

    @Override
    public int writeTimeoutMillis() {
        return writeTimeoutMillis;
    }
}

File: cache/src/main/java/com/mm/http/RealRequestInterceptorChain.kt
package com.mm.http

import com.mm.http.cache.CacheHelper
import okhttp3.*
import okhttp3.internal.EMPTY_RESPONSE
import java.io.IOException
import java.net.HttpURLConnection
import java.util.concurrent.TimeUnit

/**
 * Created by : majian
 * Date : 2021/8/25
 * 处理自定义拦截器中对request 请求更改的逻辑。
 */
class RealRequestInterceptorChain internal constructor(
    private val call: Call,
    private val cache: CacheHelper?,
    private val interceptors: List<Interceptor>,
    private val index: Int,
    private val request: Request,
    private val connectTimeoutMillis: Int,
    private val readTimeoutMillis: Int,
    private val writeTimeoutMillis: Int
) : Interceptor.Chain {
    override fun call(): Call {
        return call
    }

    override fun connectTimeoutMillis(): Int {
        return connectTimeoutMillis
    }

    override fun connection(): Connection? {
        return null
    }

    @Throws(IOException::class)
    override fun proceed(request: Request): Response {
        // Call the next interceptor in the chain.
        if (index < interceptors.size) {
            val next = copy(
                index + 1,
                request,
                connectTimeoutMillis,
                readTimeoutMillis,
                writeTimeoutMillis
            )
            val interceptor = interceptors[index]
            return interceptor.intercept(next)
        }
        val response = cache?.get(request)
        return response?.newBuilder()?.request(request)?.build()
            ?: Response.Builder().request(request)
                .protocol(Protocol.HTTP_1_1)
                .code(HttpURLConnection.HTTP_VERSION)
                .message("Unsatisfiable Request (only-if-cached)")
                .body(EMPTY_RESPONSE).build()
    }

    override fun readTimeoutMillis(): Int {
        return readTimeoutMillis
    }

    override fun request(): Request {
        return request
    }

    override fun withConnectTimeout(timeout: Int, unit: TimeUnit): Interceptor.Chain {
        val connectTimeoutMillis = checkDuration(timeout, unit)
        return copy(index, request, connectTimeoutMillis, readTimeoutMillis, writeTimeoutMillis)
    }

    override fun withReadTimeout(timeout: Int, unit: TimeUnit): Interceptor.Chain {
        val readTimeoutMillis = checkDuration(timeout, unit)
        return copy(index, request, connectTimeoutMillis, readTimeoutMillis, writeTimeoutMillis)
    }

    override fun withWriteTimeout(timeout: Int, unit: TimeUnit): Interceptor.Chain {
        val writeTimeoutMillis = checkDuration(timeout, unit)
        return copy(index, request, connectTimeoutMillis, readTimeoutMillis, writeTimeoutMillis)
    }

    private fun copy(
        index: Int, request: Request,
        connectTimeoutMillis: Int, readTimeoutMillis: Int, writeTimeoutMillis: Int
    ): RealRequestInterceptorChain {
        return RealRequestInterceptorChain(
            call,
            cache,
            interceptors,
            index,
            request,
            connectTimeoutMillis,
            readTimeoutMillis,
            writeTimeoutMillis
        )
    }

    private fun checkDuration(duration: Int, unit: TimeUnit?): Int {
        require(duration >= 0) { "duration < 0" }
        requireNotNull(unit) { "unit == null" }
        return unit.toMillis(duration.toLong()).toInt()
    }

    override fun writeTimeoutMillis(): Int {
        return writeTimeoutMillis
    }
}